### PR TITLE
[PoC] New build plan format and generic phases

### PR DESCRIPTION
**Pull request by coffee-cup**
This PR implements a new build plan format that
- Enables providers to define more phase types besides setup, install, build, and start
- Simplifies the generation of Dockerfile snippets and writing of supporting app files (nix expressions)
- Paves the way so we can better support matching against multiple providers (#237)

_Note: a lot of this PR is still a WIP and it is not 1:1 feature complete with the previous version. Some things are still a bit messy. I am opening this up early for feedback and discussion on the higher-level ideas. The "New" prefix is also just for simplifying the implementation of this PR and will be removed in the final version._

# Generic and consistent phases

Previously the build plan was comprised of four phases
- Setup (install packages)
- Install (install language deps)
- Build (build the app)
- Start (run the app when the container starts)

These phases were set in stone and hard coded. The implementation for each phase was also completely different, but very similar to each other. This made updating phases a tedious and confusing task.

This PR implements two new phase types: `Phase` and `StartPhase`. 

The generic `Phase` type can be used for any step that occurs during the creation of the image during the build. This would encapsulate the current setup, install, and build phases. Going from the phase to a dockerfile or nix expression is the same for each setup/build/install without needing a separate custom implementation for each.

The phase can also "depend on" another phase. When creating the image we will topologically sort these phases and execute them in the correct order.

```rust
#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct NewPhase {
    pub name: String,

    #[serde(rename = "dependsOn")]
    pub depends_on: Option<Vec<String>>,

    #[serde(rename = "nixPackages")]
    pub nix_pkgs: Option<Vec<Pkg>>,

    #[serde(rename = "nixLibraries")]
    pub nix_libraries: Option<Vec<String>>,

    #[serde(rename = "nixpacksArchive")]
    pub nixpacks_archive: Option<String>,

    #[serde(rename = "aptPackages")]
    pub apt_pkgs: Option<Vec<String>>,

    #[serde(rename = "commands")]
    pub cmds: Option<Vec<String>>,

    #[serde(rename = "onlyIncludeFiles")]
    pub only_include_files: Option<Vec<String>>,

    #[serde(rename = "cacheDirectories")]
    pub cache_directories: Option<Vec<String>>,
}
```

The `StartPhase` is slightly different since this specifies how the container will start.

```rust
#[serde_with::skip_serializing_none]
#[derive(Serialize, Deserialize, Default, Clone, Debug)]
pub struct NewStartPhase {
    pub cmd: Option<String>,

    #[serde(rename = "runImage")]
    pub run_image: Option<String>,

    #[serde(rename = "onlyIncludeFiles")]
    pub only_include_files: Option<Vec<String>>,
}
```

# Provider Implementation

With these new phases comes a new way to create providers. Instead of Nixpacks calling into each provider to get each phase, the provider returns a full build plan. For example, the deno provider.

```rust
fn get_build_plan(&self, app: &App, _env: &Environment) -> Result<Option<NewBuildPlan>> {
    let mut plan = NewBuildPlan::default();

    let mut setup_phase = NewPhase::new("setup");
    setup_phase.add_nix_pkgs(vec![Pkg::new("deno")]);
    plan.add_phase(setup_phase);

    if let Some(start_file) = DenoProvider::get_start_file(app)? {
        let mut build_phase = NewPhase::new("build");
        build_phase.depends_on_phase("setup");

        build_phase.add_cmd(format!(
            "deno cache {}",
            start_file
                .to_str()
                .context("Failed to convert start_file to string")?
        ));

        plan.add_phase(build_phase);
    };

    if let Some(start_cmd) = DenoProvider::get_start_cmd(app)? {
        let start_phase = NewStartPhase::new(start_cmd);
        plan.add_start_phase(start_phase);
    }

    let build_plan = NewBuildPlan::default();
    Ok(Some(build_plan))
}
```

For more complicated providers this can obviously be cleaned up and split out to `get_build_phase` functions.

I am also thinking of a few ways to improve ergonomics with macros but that can come after the plan format is solidified.

_Note: we can keep the current implementation of the providers for now and just apply a conversion from old format to new format._

# Output generation

As part of the image creation process, a few files need to be generated and saved to disk. This include things such as static assets, nix expressions, and the Dockerfile.

Previously that was all shoved into a single file and was hard to parse. In this PR each type is responsible for generating code for the Dockerfile and writing supporting files. The result is much cleaner image creation

```rust
let dockerfile = plan
    .generate_dockerfile(&self.options, env)
    .context("Generating Dockerfile for plan")?;

self.write_app(app_src, dest).context("Writing app")?;
self.write_dockerfile(dockerfile, dest)
    .context("Writing Dockerfile")?;
plan.write_supporting_files(&self.options, env, dest)
    .context("Writing supporting files")?;

let mut docker_build_cmd = self.get_docker_build_cmd(plan, name.as_str(), dest)?;

// Execute docker build
let build_result = docker_build_cmd.spawn()?.wait().context("Building image")?;
if !build_result.success() {
  bail!("Docker build failed")
}
```

# New build plan format

And finally, the new build plan format. I think this format will lend itself well to using multiple providers to build because the plans can just be merged together to make a new build plan. It was also possible with the previous format, but this is much more flexible.

```rust
#[serde_with::skip_serializing_none]
#[derive(Debug, Serialize, Deserialize, Default, Clone)]
pub struct NewBuildPlan {
    #[serde(rename = "nixpacksVersion")]
    pub nixpacks_version: Option<String>,

    pub variables: Option<EnvironmentVariables>,

    #[serde(rename = "staticAssets")]
    pub static_assets: Option<StaticAssets>,

    pub phases: Vec<NewPhase>,

    #[serde(rename = "startPhase")]
    pub start_phase: Option<NewStartPhase>,
}
```

### CODE REVIEW

**Comment by aarnphm** [comment]
Is it possible to define system packages per architecture in this new build plan proposal?

**Comment by coffee-cup** [comment]
> Is it possible to define system packages per architecture in this new build plan proposal?

What do you mean? Like something that the providers can do?

That is a separate issue/feature and can be added after with this build plan.

File: src/nixpacks/builder/docker/dockerfile_generation.rs [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+        let dockerfile = formatdoc! {"
+            FROM {base_image}
+            WORKDIR {APP_DIR}
+
+            {dockerfile_phases_str}
+
+            {start_phase_str}
+        "};
+
```
Comment (AhmedMozaly) [inline src/nixpacks/builder/docker/dockerfile_generation.rs:102]: Could this support docker multi-stage build? (one stage for creating artifact & other stage for running the artifact)
